### ISSUE TITLE
Preserve Extractor results in their original form

### ISSUE THREAD

**Issue by tyarkoni** [tyarkoni/pliers#223]
Currently, we convert the result produced by every `Extractor` to a standardized representation inside the `transform` call--i.e., by constructing an `ExtractorResult` object that already has its values as an array-like object. This makes it easy to combine `ExtractorResults` from different Stims and/or Extractors (e.g., via `merge_results`), but it's problematic in that it means users who expect to get back data in a familiar form have no way of reconstructing the original result.

For example, suppose someone already has existing code to work with results from the Google API, and just wants to use pliers as an easier way to get those results. This kind of thing is probably not going to be a terribly uncommon use case, but we don't support it right now. Instead, we force users to work with new, and likely unfamiliar, representations.

The proposed solution is to let `Extractor` classes pass whatever the wrapped tool would normally return to the user as the first argument (i.e., `data`) to the `ExtractorResult` init. In other words, instead of being something array-like, `data` can be anything at all. This makes it easy for users to, e.g., loop over a list of `ExtractorResult` objects and work with the data in a form native to the underlying tool. BUT to make sure that we can still do all the merging and DataFrame manipulations, we allow `Extractor` classes to optionally implement a `to_df()` class method that take the `data` object (or even an entire `ExtractorResult` instance and return a pandas `DataFrame`. Since `ExtractorResult` instances always store references to the originating `Extractor`, we can then call the Extractor's `.to_df()` method inside `ExtractorResult.to_df()`.

I think this gives us the best of both worlds: retention of the original feature data, regardless of representation, and a way to access the same data in a standardized form. It also gives us a lot of flexibility to take formatting arguments in the `Extractor.to_df()` method--for example, the `handle_annotations` argument in the Google extractors could potentially move to the `to_df()` call, which allows us to keep all data returned from Google, and decide how to format it only when we're actually about to do something with it.

**Comment by qmac** [comment]
Some thoughts:
- Not a huge fan of moving away from standardized output as the default
- We'd have to rework our logging/reproducibility strategies if we go with an `Extractor` level `to_df()`, as any parsing arguments that modify output format would have to be logged/part of identifying stim-->result uniqueness.
- Assuming outputs are still `ExtractorResult` objects, we could just store everything in there as attributes (`raw`, `data`, `values`, or whatever). 
- On the other hand, I do like the idea of abstracting output parsing away from `_extract` methods.

**Comment by tyarkoni** [comment]
I agree with points 1 and 3. Not sure about 2... I think it's reasonable to view the output of a `to_df()` call as not really part of the reproducible workflow, since that's essentially just output formatting. (I.e., the expectation should be that `to_df` is never going to do anything to substantively change values).

Practically, I think having optional `to_df` and `to_array` class methods in `Extractor` classes will save us a *lot* of trouble. For example, the precipitating reason for opening this issue was that I've just implemented extractors for the face_recognition package, and the format they return the data in makes it really hard to initialize a `DataFrame` properly inside the generic `ExtractorResult.to_df()` method (for reasons not worth getting into). It should really be the concern of the `Extractor` to determine how a raw result gets turned into a numpy array and/or pandas DF, in my view.

I've actually got a working prototype of this right now that's fully backward-compatible. It works by adding an optional `raw` argument to the `ExtractorResult` initializer that just stores the original object returned by the underlying extractor tool. Inside the init, we check to see if `raw is not None` and `self.transformer` has an implemented `to_df` method; if `True`, then `self.data` is generated by calling `self.transformer.to_df(raw)`; otherwise it just uses the `data` argument passed to init. This means we don't have to change any of the existing `Extractor` classes, but new classes can take advantage of the additional flexibility.

I do think that if this pattern ends up working well, we'll probably want to break the API in some future release and pass *only* the `raw` values to the `ExtractorResult` init, and make generation of the standardized result via an `Extractor` class method like `to_df` mandatory. Again, I like that approach because it nicely separates concerns. But for now, let's try it out with the non-breaking change and see how it goes. At the very least, it will solve my face_recognition problem.
